TableFactor for DiscreteFactorGraph.

ywkim0606 opened the pull request:
1. TableFactor implementation using Eigen's SparseVector.
    - lazy_cp stands for [Lazy Cartesian Product](http://phrogz.net/lazy-cartesian-product#:~:text=Lazy%20Cartesian%20Product%3A%20Iterating%20combinations,without%20calculating%20them%20all%20first.)
    - operator*(const TableFactor& f) can be implemented more elegantly using custom interator.
3. Unittests for TableFactor.
4. Added tests in testDiscreteFactorGraph to verify if inference still works.
    - Note that TableFactor gets converted to DecisionTreeFactor.
Can be useful if the factor is very sparse (CSP).

dellaert commented on gtsam/discrete/tests/testDiscreteFactorGraph.cpp:
```diff
@@ -54,6 +55,31 @@ TEST_UNSAFE(DiscreteFactorGraph, debugScheduler) {
 }
 
+/* ************************************************************************* */
+TEST_UNSAFE(DiscreteFactorGraph, debugSchedulerTable) {
```
remove UNSAFE

dellaert commented on gtsam/discrete/tests/testDiscreteFactorGraph.cpp:
```diff
@@ -100,6 +126,57 @@ TEST_UNSAFE( DiscreteFactorGraph, DiscreteFactorGraphEvaluationTest) {
 
+/* ************************************************************************* */
+/// Test the () operator of DiscreteFactorGraph
+TEST_UNSAFE(DiscreteFactorGraph, DiscreteFactorGraphEvaluationTestTable) {
```
remove UNSAFE

dellaert commented on gtsam/discrete/TableFactor.h:
```diff
@@ -0,0 +1,383 @@
+  typedef DiscreteFactor Base;  ///< Typedef to base class
+  typedef boost::shared_ptr<TableFactor> shared_ptr;
+
+  /// @}
```
remove this line. This ends a section but there is no section defined yet.

dellaert commented on gtsam/discrete/TableFactor.h:
```diff
@@ -0,0 +1,383 @@
+template <>
+struct traits<TableFactor> : public Testable<TableFactor> {};
+
+}  // namespace gtsam
```
add newline.

dellaert reviewed (changes requested):
Good PR. Please add a .cpp file and move all implementations there. Then I'll do a thorough review on .h file. Some methods are public that i don't think should be public (lazy_cp), but I'll take a look at that later.
Finally, it would be nice to get some timing results vs. DT version.

dellaert commented:
Sorry, was at ICRA - what's the status/plan on this?

ywkim0606 commented:
> Sorry, was at ICRA - what's the status/plan on this?
Status:
- Rewrote max function which eliminates variables according to the ordering given. (needs review)
- Implemented maxAssignemnt function which returns the assignments of the maximum value. 
- Both has been unittested
Plan:
- Create a new branch to edit DiscreteFactorGraph. On this branch, DiscreteFactorGraph still uses DecisionTreeFactor to optimize.
- Comparing timing results with DT on classical planning task.

dellaert commented:
- Why the new max functions? Not questioning just curious what the issue was.
- Will review, but in meantime check what CI failure is on windows? Seems related to your code.

dellaert commented on gtsam/discrete/TableFactor.cpp:
```diff
@@ -0,0 +1,340 @@
+}
+
+/* ************************************************************************ */
+TableFactor TableFactor::max(size_t nrFrontals) const {
```
can you not call the method below?

ywkim0606 replied:
This function would max by eliminating the variables in incremental order without having to explicitly give an ordering. I thought this was why we had a separate function in DecisionTreeFactor so I added the same feature in TableFactor too. We can call the method below by explicitly giving the ordering in incremental order.

dellaert commented on gtsam/discrete/TableFactor.cpp:
```diff
@@ -0,0 +1,340 @@
+                       "keys %d, nr.keys=%d") %
+         frontalKeys.size() % size())
+            .str());
+  TableFactor result(*this);
```
What is happening here? How is eliminate a max ? Don't just explain in reply, add a comment.

ywkim0606 replied:
comment added

dellaert commented on gtsam/discrete/TableFactor.cpp:
```diff
@@ -0,0 +1,340 @@
+}
+
+/* ************************************************************************ */
+TableFactor TableFactor::eliminate(const Key key) const {
```
We prefer calling "up" i.e. in this case eliminate should be defined (and declared in .h) *before* max.

dellaert replied:
Add some comments on how this function works.

dellaert commented on gtsam/discrete/TableFactor.cpp:
```diff
@@ -0,0 +1,340 @@
+TableFactor TableFactor::fromDecisionTreeFactor(
+    const DecisionTreeFactor& f) const {
+  DiscreteKeys tree_dkeys = f.discreteKeys();
+  std::vector<std::pair<DiscreteValues, double>> enumerated = f.enumerate();
```
auto?

dellaert commented on gtsam/discrete/TableFactor.cpp:
```diff
@@ -0,0 +1,340 @@
+/* ************************************************************************ */
+TableFactor TableFactor::fromDecisionTreeFactor(
+    const DecisionTreeFactor& f) const {
+  DiscreteKeys tree_dkeys = f.discreteKeys();
```
declare closer to where you use this.

dellaert commented on gtsam/discrete/TableFactor.cpp:
```diff
@@ -0,0 +1,340 @@
+DiscreteValues TableFactor::maxAssignment() const {
+  DiscreteValues assignments;
+  Eigen::Index nnz = sparse_table_.nonZeros();
+  Eigen::Index rowIdx;
```
`rowIdx` -> `maxRowIndex`

dellaert commented on gtsam/discrete/TableFactor.cpp:
```diff
@@ -0,0 +1,340 @@
+}
+
+/* ************************************************************************ */
+std::vector<DiscreteValues> TableFactor::project(
```
add some comments on how this function will proceed

dellaert commented on gtsam/discrete/TableFactor.h:
```diff
@@ -0,0 +1,199 @@
+              const Eigen::SparseVector<double>& table);
+
+  /** Convert table to SparseVector */
+  static Eigen::SparseVector<double> Convert(const std::vector<double>& table);
```
These should be private I think

dellaert commented on gtsam/discrete/TableFactor.h:
```diff
@@ -0,0 +1,199 @@
+  /// Find index for given assignment of values
+  size_t findIndex(const DiscreteValues& assignment) const;
+
+  /// Find indicies for given partial assignment of values
```
I do not understand the comment, or why this is eliminate.

ywkim0606 replied:
Sorry, that was a depreciated comment. Will be updated on the upcoming commit.

dellaert commented on gtsam/discrete/TableFactor.h:
```diff
@@ -0,0 +1,199 @@
+   */
+  std::string markdown(const KeyFormatter& keyFormatter = DefaultKeyFormatter,
+                       const Names& names = {}) const override {
+    return "";
```
Move to cpp file and fill in :-)

dellaert commented on gtsam/discrete/TableFactor.h:
```diff
@@ -0,0 +1,199 @@
+   */
+  std::string html(const KeyFormatter& keyFormatter = DefaultKeyFormatter,
+                   const Names& names = {}) const override {
+    return "";
```
same

dellaert reviewed (changes requested):
Not quite there yet.

ywkim0606 commented:
> * Why the new max functions? Not questioning just curious what the issue was.
> * Will review, but in meantime check what CI failure is on windows? Seems related to your code.

I miss interpreted the comment in DecisionTreeFactor "Create new factor by maximizing over all values with the same separator". What I did was maximize over a single key looking at just its cardinality.

dellaert commented on gtsam/discrete/TableFactor.h:
```diff
@@ -0,0 +1,195 @@
+  shared_ptr sum(size_t nrFrontals) const;
+
+  /// Create new TableFactor where the input key is eliminated
+  TableFactor eliminate(const Key key) const;
```
just `(Key key)`, as is passed by value hence implied.

dellaert commented on gtsam/discrete/TableFactor.h:
```diff
@@ -0,0 +1,195 @@
+  /// Create new factor by summing all values with the same separator values
+  shared_ptr sum(size_t nrFrontals) const;
+
+  /// Create new TableFactor where the input key is eliminated
```
`/// Create new TableFactor where the variable with given key is eliminated.`

But I don't understand this method! Whenever we *eliminate* there is a typically *both* a conditional and a factor that is computed. In addition, elimination functions are typically free-standing functions, see, e.g., `EliminateDiscrete` and `EliminateForMPE` in `DiscreteFactorGraph.cpp`. 

If the above is not clear, I think we might have to meet about this.

dellaert commented on gtsam/discrete/TableFactor.h:
```diff
@@ -0,0 +1,195 @@
+  TableFactor fromDecisionTreeFactor(const DecisionTreeFactor& f) const;
+
+  /// Finds value for the key at index
+  size_t lazy_cp(Key target_key, size_t index) const;
```
Naming convention (`lazyCartesianProduct` ?) and should this not be private?

ywkim0606 replied:
We can make this private but we won't be able to run unittest on this function anymore. However, since the function has been tested already I'll make it private.

dellaert commented on gtsam/discrete/TableFactor.h:
```diff
@@ -0,0 +1,195 @@
+  size_t findIndex(const DiscreteValues& assignment) const;
+
+  /// Create union of keys
+  DiscreteKeys unionKeys(const TableFactor& f) const;
```
private?

dellaert commented on gtsam/discrete/TableFactor.h:
```diff
@@ -0,0 +1,195 @@
+  /// Create union of keys
+  DiscreteKeys unionKeys(const TableFactor& f) const;
+
+  /// Project assignment that agrees with the given assignment
```
Add example here. I suggested this function but from the comment I can for the life of me not understand what this method is for or what it produces.

dellaert commented on gtsam/discrete/TableFactor.cpp:
```diff
@@ -0,0 +1,415 @@
+  new_sparse_table_.reserve(sparse_table_.nonZeros() *
+                            (cardinality / sparse_table_.size()));
+  TableFactor multiplied_f(keys, new_sparse_table_);
+  for (Eigen::SparseVector<double>::InnerIterator it(f.sparse_table_); it;
```
maybe say `using Eigen::SparseVector<double>::InnerIterator` at top of this file so you can just use `InnerIterator`....

ywkim0606 replied:
"Error: a class qualified name is not allowed". Instead, I `typedef` it: "typedef Eigen::SparseVector<double>::InnerIterator sparse_it;"

dellaert commented on gtsam/discrete/tests/testTableFactor.cpp:
```diff
@@ -0,0 +1,323 @@
+// Check markdown representation looks as expected.
+TEST(TableFactor, markdown) {
+  DiscreteKey A(12, 3), B(5, 2);
+  TableFactor f(A & B, "1 2  3 4  5 6");
```
How about making it sparse :-)

dellaert reviewed (changes requested):
Awesome progress. Very important class, so let's iterate once more. Especially confusion around eliminate and project at the moment.

ywkim0606 commented on gtsam/discrete/TableFactor.h:
```diff
@@ -0,0 +1,195 @@
+  TableFactor(const DiscreteKeys& keys, const std::string& table)
+      : TableFactor(keys, Convert(table)) {}
+
+  /// Single-key specialization
```
This constructor is written to construct a factor for one variable (DiscreteKey instead of DiscreteKeys). Should I rephrase to "Constructor for single-key"?

dellaert commented:
@ywkim0606  What's the status here?

ywkim0606 commented:
> @ywkim0606 What's the status here?

I'll schedule a meeting this week regarding elimination.
